Split lens and source photo-z catalogs

joezuntz opened the pull request:
This removes the assumption in the PZ parts of the pipeline that the source and lens catalogs come from the same underlying catalog.

This is in preparation for Metadetect where this will no longer be true.  After this I have to do something similar for the mapping stages and TXDiagnostics.

To cope with non-ugrizy data in RAIL this also updates to the newest version of that submodule.

chihway commented on txpipe/rail_pz.py:
```diff
@@ -57,20 +65,66 @@ def run(self):
+
+    outputs = [(model_tag, PickleFile)]
+
+class PZRailTrainLensFromSource(PipelineStage):
```
this feels a little odd, maybe it's the name... like why is it not trainsourcefromlens? also, what if we only run part of 3x2, ie clustering only? (maybe that's a larger question)

joezuntz replied:
Yeah, good idea, I've added the reverse stage SourceFromLens

chihway commented on txpipe/rail_pz.py:
```diff
@@ -197,3 +270,74 @@ def write_output_chunk(self, output_file, start, end, z, pz_data):
+
+
+
+class PZRailEstimateSource(PZRailEstimateLens):
```
is there a logic whether the source or the lens is the parent class in estimate vs. train?

joezuntz replied:
There wasn't, so I've harmonized this a bit.

chihway commented on txpipe/rail_pz.py:
```diff
@@ -197,3 +270,74 @@ def write_output_chunk(self, output_file, start, end, z, pz_data):
+            nobj = f["shear/ra"].size
+        return nobj
+
+class PZRailEstimateSourceFromLens(PipelineStage):
```
same question as before

chihway commented:
looks good, just minor questions about naming, feel free to ignore if it makes no sense

joezuntz commented on examples/laptop_pipeline.yml:
```diff
@@ -6,10 +6,14 @@ stages:
       threads_per_process: 2
-    - name: PZRailEstimate
-    - name: TXPhotozStack        # stack p(z) into n(z)
+    - name: PZRailTrainLens
```
Done!

joezuntz replied:
I've also modified to speed things up by using the same estimator for the two.

joezuntz commented:
Thanks!  I've addressed your issues and also split rail off into two files in a new dir, but with no other changes.
I'll merge if the tests run okay.
